draft-pauly-add-deer.md

```diff
@@ -170,44 +170,49 @@ Encrypted Resolver name.
-address than the Unencrypted Resolver and the TLS certificate does not cover the
-Unencrypted Resolver address, the client MUST NOT use the discovered Encrypted
+Unencrypted Resolver.  Otherwise, the client MUST NOT use the discovered Encrypted
 Resolver. Additionally, the client SHOULD suppress any further queries for
 Equivalent Encrypted Resolvers using this Unencrypted Resolver for the length of
 time indicated by the SVCB record's Time to Live (TTL).
 
-If the Equivalent Encrypted Resolver and the Unencrypted Resolver share an IP
-address, clients MAY choose to opportunistically use the Encrypted Resolver even
-without this certificate check ({{opportunistic}}).
-
 ## Opportunistic Discovery {#opportunistic}
 
-There are situations where authenticated discovery of encrypted DNS
-configuration over unencrypted DNS is not possible. This includes Unencrypted
-Resolvers on non-public IP addresses whose identity cannot be confirmed using
-TLS certificates.
+If the Unencrypted Resolver's IP address is in a private range, it is considered
+a local reference identity that cannot be confirmed using TLS certificates.
+Instead, the ability to respond to queries confirms control of the address.
+The client MUST repeat the discovery process at least every 5 minutes (e.g. by
+limiting the SVCB response TTL) to confirm that the result is still valid, and
+MUST stop using any Encrypted Resolver that is no longer included.
 
 Opportunistic Privacy is defined for DoT in Section 4.1 of {{!RFC7858}} as a
 mode in which clients do not validate the name of the resolver presented in the
 certificate. A client MAY use information from the SVCB record for
 "dns://resolver.arpa" with this "opportunistic" approach (not validating the
 names presented in the SubjectAlternativeName field of the certificate) as long
-as the IP address of the Encrypted Resolver does not differ from the IP address
-of the Unencrypted Resolver, and that IP address is a private address (such as
+as the IP address of the Encrypted Resolver is also a private address (such as
 those defined in {{!RFC1918}}). This approach can be used for DoT or DoH.
 
-If the IP addresses of the Encrypted and Unencrypted Resolvers are not the same,
-or the shared IP address is not a private IP address, the client MUST NOT use
-the Encrypted Resolver opportunistically.
+In order to refer the user to an Encrypted Resolver with a public IP address, the
+Unencrypted Resolver MUST return a CNAME that aliases `_dns.resolver.arpa` to
+`_dns.$HOSTNAME`.  If the client receives such a CNAME, it MAY proceed with
+name-based discovery for `$HOSTNAME` ({{encrypted}}), subject to the above TTL
+limit.  Clients MUST NOT connect to an Encrypted Resolver on a public IP address
+without TLS authentication.
+
+Resolvers that support authenticated discovery ({{authenticated}}) can also support
+opportunistic discovery via local forwarders by publishing such a CNAME record and
+placing their SVCB records at `_dns.$HOSTNAME`.
```

Comment:
Doesn't this create a very large danger of the encrypted resolver not giving equivalent (for the various recently discussed definitions thereof) behavior to the local forwarder? E.g., if the local forwarder is providing parental-control filtering before forwarding to an upstream resolver, then the upstream encrypted resolver will "steal" resolution and unexpectedly-to-the-user disable filtering.  You could also start leaking previously-local-network-only data if the local forwarding was resolving some local-only names that could now get leaked to the external upstream resolver.

Seems we'd be in danger of a whole bunch of issues unless all the local forwarders in the world get updated to block _dns.resolver.arpa whenever they are doing something extra that the upstream resolver will not replicate.

Reply:
Thanks for the reminder; I've added a sentence reminding forwarders that they can block this domain.

This design parallels Mozilla's [canary domain](https://support.mozilla.org/en-US/kb/canary-domain-use-application-dnsnet).  That design appears have been [deployed successfully](https://datatracker.ietf.org/meeting/106/materials/slides-106-abcd-mozilla-canary-domain), and I believe it represents the working group's consensus on the best balance between improving security and ensuring local control.

One key observation from that deployment is that filtering forwarders have to be able to update their blocklist rapidly, so even old systems that can't add functionality can (and did) move quickly to add the canary domain to their list.

Reply:
Even with Firefox's experience, I'm still hesitant to start basing further solutions on such an untrusted network signal.  I recall that Mozilla has indicated in the past that they reserve the right to stop trusting or otherwise change the canary signal if they detect significant signs of abuse, and I think abuse only becomes more likely if the canary gets standardized and more clients adopt it.  I know I would feel a lot more comfortable if we could at least come up with a way to ensure the canary signal is coming from the local resolver being switched away from, rather than something else like an ISP manipulating the beyond-home network.

Also note that Firefox's deployment experience in this area so far is only for the more narrowly-scoped US-only and only affects change to offline-verified "trustworthy" resolvers.  Only partially applicable to the more general scope under consideration here.

And putting potential controversy aside, while I think it may still be reasonable to standardize a canary domain for resolvers to block to signal a request to not change resolvers, whether or not clients will trust or respect such a signal is more on the policy side of things.  So if we're going to rely on it to make this local->external upgrade safe, I think we're going to want to at least keep the same-IP option for those clients that choose not to trust the canary.

Reply:
See, the way I read this, clients can but don't have to use the CNAME method of a local forwarder referring to a public server given it's a MAY. It's an optional untrusted signal for resolvers who want to make themselves discoverable more conveniently but less securely. Clients who won't settle for opportunistic upgrade won't use this. Am I off-base?

That said, I do agree with Eric: the canary domain is not an ideal path forward. We should be focused primarily on upgrading securely without prior knowledge needed (which TRR requires).

Reply:
@ericorth I don't think DEER is subject to any of the canary abuse considerations that affect Mozilla's TRR arrangement.  Although DEER bears some resemblance to Mozilla's canary, it is much less problematic: each server is simply speaking about itself, with each forwarder having precedence over its destination.

I don't believe there is any loss of security in this aspect of the PR.  What's potentially lost is a particular edge-case of legacy compatibility.  I believe that case is narrow, and easily remedied.

@mstojens Indeed, the opportunistic modes of DEER are all independently optional, so clients are always free to apply additional restrictions and abandon the upgrade if they are not met.

Reply:
@ericorth I added a new section on the legacy device question, which also gets at the question of the range of allowable client behaviors.  Does this work for you?